## Week 2 - Homework

Jorge Vladimir Abrego Arevalo

In [1]:
import os

### Q1. Install the package
What's the version that you have?

In [2]:
!mlflow --version

mlflow, version 2.3.2


### Q2. Download and preprocess the data
Executing command

In [3]:
!python preprocess_data.py --raw_data_path ./data --dest_path ./output

In [4]:
for path in os.scandir('./output'):
    if path.is_file():
        print(path.name)

dv.pkl
test.pkl
train.pkl
val.pkl


In [5]:
file_stats = os.stat('./output/dv.pkl')
print(f'File Size in Kilobytes is {round(file_stats.st_size / (1024),2)}')

File Size in Kilobytes is 150.06


### Q3. Train a model with autolog

What is the value of the `max_depth` parameter:

>  rf = RandomForestRegressor(max_depth=10, random_state=0)

In [6]:
!python train.py

2023/05/31 23:40:59 INFO mlflow.store.db.utils: Creating initial MLflow database tables...
2023/05/31 23:40:59 INFO mlflow.store.db.utils: Updating database tables
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> 451aebb31d03, add metric step
INFO  [alembic.runtime.migration] Running upgrade 451aebb31d03 -> 90e64c465722, migrate user column to tags
INFO  [alembic.runtime.migration] Running upgrade 90e64c465722 -> 181f10493468, allow nulls for metric values
INFO  [alembic.runtime.migration] Running upgrade 181f10493468 -> df50e92ffc5e, Add Experiment Tags Table
INFO  [alembic.runtime.migration] Running upgrade df50e92ffc5e -> 7ac759974ad8, Update run tags with larger limit
INFO  [alembic.runtime.migration] Running upgrade 7ac759974ad8 -> 89d4b8295536, create latest metrics table
INFO  [89d4b8295536_create_latest_metrics_table_py] Migration complete!
INFO  

### Launch the tracking server locally for MLflow

For this task, run:

> mlflow ui --backend-store-uri sqlite:///mlflow.db --artifacts-destination ./artifacts

### Q4. Tune model hyperparameters

In [8]:
!python hpo.py

2023/05/31 23:43:34 INFO mlflow.tracking.fluent: Experiment with name 'random-forest-hyperopt' does not exist. Creating a new experiment.
[I 2023-05-31 23:43:34,986] A new study created in memory with name: no-name-70fef869-e7b4-455a-9edb-003723f06b1c
[I 2023-05-31 23:43:36,832] Trial 0 finished with value: 2.451379690825458 and parameters: {'n_estimators': 25, 'max_depth': 20, 'min_samples_split': 8, 'min_samples_leaf': 3}. Best is trial 0 with value: 2.451379690825458.
[I 2023-05-31 23:43:37,080] Trial 1 finished with value: 2.4667366020368333 and parameters: {'n_estimators': 16, 'max_depth': 4, 'min_samples_split': 2, 'min_samples_leaf': 4}. Best is trial 0 with value: 2.451379690825458.
[I 2023-05-31 23:43:38,839] Trial 2 finished with value: 2.449827329704216 and parameters: {'n_estimators': 34, 'max_depth': 15, 'min_samples_split': 2, 'min_samples_leaf': 4}. Best is trial 2 with value: 2.449827329704216.
[I 2023-05-31 23:43:39,459] Trial 3 finished with value: 2.460983516558473 a

What's the best validation RMSE that you got?

<img src="images/Question 4 - RMSE with hyperopt.png" alt="Best RMSE" />

The best validation RMSE is 2.45

### Q5. Promote the best model to the model registry

In [9]:
!python register_model.py

2023/05/31 23:51:19 INFO mlflow.tracking.fluent: Experiment with name 'random-forest-best-models' does not exist. Creating a new experiment.
2023/05/31 23:51:23 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "C:\Anaconda3\envs\mlops\lib\site-packages\_distutils_hack\__init__.py:33: UserWarning: Setuptools is replacing distutils."
Successfully registered model 'RFR-best-model'.
2023/05/31 23:51:38 INFO mlflow.tracking._model_registry.client: Waiting up to 300 seconds for model version to finish creation. Model name: RFR-best-model, version 1
Created version '1' of model 'RFR-best-model'.


What is the test RMSE of the best model?

<img src="images/Question 5 - RMSE on test.png" alt="RMSE on test" />

The test RMSE of best model is 2.291

### Q6. Model metadata

Now explore your best model in the model registry using UI. What information does the model registry contain about each model?

<img src="images/Question 6 - Information that model registry contains about each model.png" alt="RMSE on test" />

Version Number